test/Dialect/Torch/simplify-dtype-calculations.mlir

```diff
@@ -0,0 +1,284 @@
+    %arg1_as_tensor = torch.prim.NumToTensor.Scalar %arg1 : !torch.int -> !torch.tensor<[],si64>
+    %si64_dtype = torch.prim.dtype %arg1_as_tensor : !torch.tensor<[],si64> -> !torch.int
+    %dtypes = torch.prim.ListConstruct %f64_dtype, %si64_dtype : (!torch.int, !torch.int) -> !torch.list<int>
+    %7 = torch.promote_dtypes %ranks, %dtypes : (!torch.list<optional<int>>, !torch.list<int>) -> !torch.int
+    torch.dtype.calculate.yield.dtypes %7 : !torch.int
+  } : !torch.number
+  return %0 : !torch.number
+}
+
+// -----
+
+// CHECK-LABEL:   func.func @promote_dtypes$scalar_scalar_same_category(
+// CHECK:             {{.*}} = torch.aten.add {{.*}} -> !torch.int
+func.func @promote_dtypes$scalar_scalar_same_category(%arg0: !torch.int, %arg1: !torch.int) -> !torch.number {
+  %none = torch.constant.none
+  %0 = torch.dtype.calculate {
+    %1 = torch.aten.add %arg0, %arg1 : !torch.int, !torch.int -> !torch.number
+    torch.dtype.calculate.yield %1 : !torch.number
+  } dtypes {
+    %ranks = torch.prim.ListConstruct %none, %none : (!torch.none, !torch.none) -> !torch.list<optional<int>>
+    %arg0_as_tensor = torch.prim.NumToTensor.Scalar %arg0 : !torch.int -> !torch.tensor<[],si64>
+    %si64_dtype = torch.prim.dtype %arg0_as_tensor : !torch.tensor<[],si64> -> !torch.int
+    %dtypes = torch.prim.ListConstruct %si64_dtype, %si64_dtype : (!torch.int, !torch.int) -> !torch.list<int>
+    %7 = torch.promote_dtypes %ranks, %dtypes : (!torch.list<optional<int>>, !torch.list<int>) -> !torch.int
+    torch.dtype.calculate.yield.dtypes %7 : !torch.int
+  } : !torch.number
+  return %0 : !torch.number
+}
+
+// -----
+
+// CHECK-LABEL:   func.func @refine_dtype$invalid_dtype_for_scalar(
+// CHECK:             {{.*}} = torch.aten.add {{.*}} -> !torch.number
+func.func @refine_dtype$invalid_dtype_for_scalar(%arg0: !torch.int, %arg1: !torch.int) -> !torch.number {
+  %none = torch.constant.none
+  %0 = torch.dtype.calculate {
+    %1 = torch.aten.add %arg0, %arg1 : !torch.int, !torch.int -> !torch.number
+    torch.dtype.calculate.yield %1 : !torch.number
+  } dtypes {
+    // dtype int for int32
+    %int3 = torch.constant.int 3
+    torch.dtype.calculate.yield.dtypes %int3 : !torch.int
+  } : !torch.number
+  return %0 : !torch.number
+}
+
+// -----
+
+// CHECK-LABEL:   func.func @refine_dtype$no_simplification
+// CHECK:             {{.*}} = torch.aten.add {{.*}} -> !torch.number
+func.func @refine_dtype$no_simplification(%arg0: !torch.int, %arg1: !torch.int, %dtype: !torch.int) -> !torch.number {
+  %none = torch.constant.none
+  %0 = torch.dtype.calculate {
+    %1 = torch.aten.add %arg0, %arg1 : !torch.int, !torch.int -> !torch.number
+    torch.dtype.calculate.yield %1 : !torch.number
+  } dtypes {
+    torch.dtype.calculate.yield.dtypes %dtype : !torch.int
+  } : !torch.number
+  return %0 : !torch.number
+}
+
+// -----
+
+// If result type is already refined (even if wrong, as is the case here), don't make any changes
+// to result type.
```

Comment:
maybe add a TODO here (or appropriately in the code) to raise an error if the inferred type is not a supertype of the already-present result type. An op like you have written here is actually statically invalid (nasal demons UB) so we probably don't want it here long-term.